### PR TITLE
SparseFunction mpi

### PR DESCRIPTION
**Pull request by mloubout**
Fixes SparseFunction with mpi.

Currently, any SparseFunction out of an operator loses the *mandatory and defining* 

map coords[i] -> data[i]

This fixes so that a give trace `data[:,i]` has its coordinates into `coords[i,:]`. This is mandatory for any operation on SparseFunction such has data residual (rec1- rec2) in FWI/RTM or just saving the result of an operator as the data just by itself doesn't mean anything

### CODE REVIEW

File: devito/types/sparse.py [hunk 1]
```diff
@@ -284,11 +293,16 @@ def _arg_values(self, **kwargs):
 
         return values
 
-    def _arg_apply(self, dataobj, alias=None):
+    def _arg_apply(self, dataobj, coordsobj, alias=None):
         key = alias if alias is not None else self
         if isinstance(key, AbstractSparseFunction):
             # Gather into `self.data`
-            key._dist_gather(self._C_as_ndarray(dataobj))
+            # Coords maye be empty if there is more ranks than coordinates
```
Comment (georgebisbas) [inline devito/types/sparse.py]: *may

File: devito/types/sparse.py [hunk 2]
```diff
@@ -826,10 +838,21 @@ def _dist_gather(self, data):
...
+            # Send out the sparse point coordinates
+            sshape, scount, sdisp, _, rcount, rdisp = self._dist_subfunc_alltoall
+            gathered = np.empty(shape=sshape, dtype=self.coordinates.dtype)
+            mpitype = MPI._typedict[np.dtype(self.coordinates.dtype).char]
+            comm.Alltoallv([coords, rcount, rdisp, mpitype],
+                           [gathered, scount, sdisp, mpitype])
+
+            self._coordinates.data._local[:] = gathered[self._dist_subfunc_gather_mask]
 
         # Note: this method "mirrors" `_dist_scatter`: a sparse point that is sent
```
Comment (georgebisbas) [inline devito/types/sparse.py]: Maybe rephrase? I am confused a bit here.

Comment (FabioLuporini) [reply devito/types/sparse.py]: I rather have another comment: @mloubout read this
"... However, the `coordinates` SubFunction values are not distributed, as this is a read-only field."

is this assumption what you're actually dropping in this PR?

File: devito/types/sparse.py [hunk 3]
```diff
@@ -826,10 +838,21 @@ def _dist_gather(self, data):
...
         comm.Alltoallv([data, rcount, rdisp, mpitype],
                        [gathered, scount, sdisp, mpitype])
-        data = gathered
         # Unpack data values so that they follow the expected storage layout
-        data = np.ascontiguousarray(np.transpose(data, self._dist_reorder_mask))
-        self._data[:] = data[self._dist_gather_mask]
+        gathered = np.ascontiguousarray(np.transpose(gathered, self._dist_reorder_mask))
+        self._data[:] = gathered[self._dist_gather_mask]
+
+        if coords is not None:
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/types/sparse.py]: Empty array is no an issue this is not for that case, this is because it can actually be None for example with `norm(rec)`  or `mmin/mmax(rec)` where the coordinates are not in args or kwargs and therefore `args[p.coordinates.name]` is None.

**Comment by mloubout** [comment]
>rec.coords is a read-only field, so I don't understand why is data gathering necessary upon returning from an Operator

As I said in the description, the whole point of a receiver is "what is the value of the wavefield at xi, yi,zi". Currently there is 0 way to know so rec is just an array with zero significance and unusable as it is just a bunch of data with no information about what grid point it corresponds too. A receiver is  a map `position => value` and the current way just kills this mapping at operator return. People won't care how fast/well it runs if the results are not usable especially people like O&G that only want proper data.

File: devito/operator.py [hunk 4]
```diff
@@ -401,7 +401,10 @@ def _prepare_arguments(self, **kwargs):
     def _postprocess_arguments(self, args, **kwargs):
         """Process runtime arguments upon returning from ``.apply()``."""
         for p in self.parameters:
-            p._arg_apply(args[p.name], kwargs.get(p.name))
+            try:
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/operator.py]: No it is necessary as Precomputed sparse function do not have a coordinates attribute but are SparseFunction

[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/operator.py]: ugh....painful...

File: devito/types/sparse.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/types/sparse.py]: ah ! fair point

**Comment by FabioLuporini** [comment]
> People won't care how fast/well it runs if the results are not usable especially people like O&G that only want proper data.

I'd have never imagined this !

> A receiver is a map position => value and the current way just kills this mapping at operator return

it's that _and the current way just kills this mapping at operator return_

I totally get the point of this PR; what I still have to figure out is why this killing as the way I thought it behaved there should be no killing at all. I'm doing another pass at the code (not just the PR) now

**Comment by FabioLuporini** [comment]
Merged.
